File: 07/Java/TripleExpression/CheckedNegate.java
package TripleExpression;

public class CheckedNegate implements TripleExpression {
	private TripleExpression value;
    public CheckedNegate(TripleExpression val) {
        value = val;
    }

    private void check(int a) throws Exception {
        if (a <= Integer.MIN_VALUE) {
            throw new OverflowException();
        }
    }

    public int evaluate(int x, int y, int z) throws Exception {
        int a = value.evaluate(x, y, z);
        check(a);
        return -a;
    }
}

File: 07/CSharp/TripleExpressionApp/TripleExpressionApp/CheckedNegate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    class CheckedNegate : ITripleExpression
    {
        private ITripleExpression value;

        public CheckedNegate(ITripleExpression value)
        {
            this.value = value;
        }

        protected void Check(double a)
        {
            if(a <= double.MinValue)
            {
                throw new OverflowException();
            }
        }

        public double Evaluate(double x, double y, double z)
        {
            double a = value.Evaluate(x, y, z);
            Check(a);
            return -a;
        }
    }
}
